### PR TITLE
SDAF calculate subnets

### PR DESCRIPTION
**Pull request by lpalovsky**
Current SDAF test use static network space settings which prevents running 
multiple concurrent tests. This PR adds functionality to create and assign 
unique addess space for each test using 'lease' file and a set of rules inside Azure container 
named 'network-spaces'. 

- Related ticket: https://jira.suse.com/browse/TEAM-9365
- Verification run: 

**3 parallel tests:**
Start condition:
2 lease files older than 7200s:
192.168.39.192
192.168.42.128

**Test 1:** 
Found both lease files available
https://openqaworker15.qa.suse.cz/tests/298290#step/deploy_workload_zone/75
Picked 192.168.39.192
https://openqaworker15.qa.suse.cz/tests/298290#step/deploy_workload_zone/95
- modification time updated

**Test 2:** 
https://openqaworker15.qa.suse.cz/tests/298288#step/deploy_workload_zone/75
192.168.42.128 - is taken as 192.168.39.192 is already used by Test 1
https://openqaworker15.qa.suse.cz/tests/298288#step/deploy_workload_zone/95

**Test 3.**
No file older than 7200s available - new file created
https://openqaworker15.qa.suse.cz/tests/298292#step/deploy_workload_zone/76

### CODE REVIEW

File: lib/sles4sap/sap_deployment_automation_framework/deployment.pm [hunk 1]
```diff
@@ -421,63 +419,6 @@ sub serial_console_diag_banner {
     wait_serial(qr/:~|#|>/, timeout => 5, quiet => 1);
 }
 
-=head2 prepare_tfvars_file
```
Comment (lpalovsky) [inline lib/sles4sap/sap_deployment_automation_framework/deployment.pm:424]: Moved to separate library

File: lib/sles4sap/azure_cli.pm [hunk 2]
```diff
@@ -12,7 +12,7 @@ use testapi;
 use Carp qw(croak);
 use Exporter qw(import);
 use Mojo::JSON qw(decode_json);
-
+use qesapdeployment qw(qesap_az_validate_uuid_pattern);
```
Comment (mpagot) [inline lib/sles4sap/azure_cli.pm:15]: it creates a dependency hierarchy that does not sound right to me: what about moving the functio nin the more generic az lib like this

https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20288

File: lib/sles4sap/azure_cli.pm [hunk 3]
```diff
@@ -1374,3 +1378,163 @@ sub az_resource_delete {
...
+
+=over 4
+
+=item B<container_name> Existing storage container name.
+
+=item B<storage_account_name> Storage account name.
+
+=item B<file> File to upload.
+
+=item B<timeout> Timeout for az command.
```
Comment (mpagot) [inline lib/sles4sap/azure_cli.pm:1399]: Other functions in this file are indicating the default value in the documentation

File: lib/sles4sap/azure_cli.pm [hunk 4]
```diff
@@ -1374,3 +1378,163 @@ sub az_resource_delete {
...
+    $args{lease_duration} //= '-1';    # -1 = infinite lease
+
+    my $az_cmd = join(' ',
+        'az storage blob lease acquire',
+        '--only-show-errors',
+        "--container-name $args{container_name}",
+        "--account-name $args{storage_account_name}",
+        "--blob-name $args{blob_name}",
+        "--lease-duration $args{lease_duration}",
+        '--output tsv'    # Json output won't work here.
```
Comment (mpagot) [inline lib/sles4sap/azure_cli.pm:1458]: could you please explain and annotate why not?

File: lib/sles4sap/azure_cli.pm [hunk 5]
```diff
@@ -1374,3 +1378,163 @@ sub az_resource_delete {
...
+    );
+
+    my $lease_id = script_output($az_cmd, proceed_on_failure => 1);
+    return ($lease_id) if qesap_az_validate_uuid_pattern($lease_id);    # Return only valid output.
```
Comment (mpagot) [inline lib/sles4sap/azure_cli.pm:1462]: could you please move the documentatio about the return code in the function documentation. How it is expected to works in case of error, or better to say how the caller should handle the "empty return" in case of error?

Comment (mpagot) [reply lib/sles4sap/azure_cli.pm:1462]: I see, like this

```
$lease_id = az_storage_blob_lease_acquire(%arguments);

    # Return unless there file lease was successful
    return unless $lease_id;
```

File: lib/sles4sap/azure_cli.pm [hunk 6]
```diff
@@ -1374,3 +1378,163 @@ sub az_resource_delete {
...
+        "--container-name $args{container_name}",
+        "--account-name $args{storage_account_name}",
+        "--query \"$args{query}\"",
+        '--output json'
+    );
+
+    return decode_json(script_output($az_cmd));
+}
+
+=head2 az_storage_blob_update
```
Comment (mpagot) [inline lib/sles4sap/azure_cli.pm:1504]: once merged we can think about using it in `qesap_az_list_container_files`

File: lib/sles4sap/azure_cli.pm [hunk 7]
```diff
@@ -1374,3 +1378,163 @@ sub az_resource_delete {
...
+=back
+=cut
+
+sub az_storage_blob_update {
+    my (%args) = @_;
+    foreach ('container_name', 'account_name', 'name') {
+        croak "Missing mandatory argument: '$_'" unless $args{$_};
+    }
+
+    my @az_cmd = (' ',
```
Comment (mpagot) [inline lib/sles4sap/azure_cli.pm:1529]: initial space is not needed

File: lib/sles4sap/azure_cli.pm [hunk 8]
```diff
@@ -1374,3 +1378,163 @@ sub az_resource_delete {
...
+        'az storage blob update',
+        '--only-show-errors',
+        "--container-name $args{container_name}",
```
Comment (mpagot) [inline lib/sles4sap/azure_cli.pm:1532]:
```suggestion
        '--container-name', $args{container_name},
```

File: lib/sles4sap/azure_cli.pm [hunk 9]
```diff
@@ -1374,3 +1378,163 @@ sub az_resource_delete {
...
+    );
+
+    return decode_json(script_output($az_cmd));
+}
+
+=head2 az_storage_blob_update
+
+    az_storage_blob_update(container_name=>'container', account_name=>'stuff', name='blobby' [, lease_id=42]);
+
+Update properties of storage blob.
```
Comment (mpagot) [inline lib/sles4sap/azure_cli.pm:1508]: say something about the return. Seems you are intentionally using `script_run` instead of `assert_script_run` to return the exit code and let the caller to decide. It is fine but makes sense to document it.

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm
Comment (mpagot) [inline lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm:1]: This file deserve its own PR

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        sap_system => data_url('sles4sap/sap_deployment_automation_framework/SAP_SYSTEM.tfvars'),
+        workload_zone => data_url('sles4sap/sap_deployment_automation_framework/WORKLOAD_ZONE.tfvars'),
+        library => data_url('sles4sap/sap_deployment_automation_framework/LIBRARY.tfvars')
+    );
+
+    # replace default vnet name with shorter one to avoid naming restrictions
+    set_workload_vnet_name();
+
+    my $tfvars_file = get_os_variable($tfvars_os_variable{$args{deployment_type}});
+    my $retrieve_tfvars_cmd = join(' ', 'curl', '-v', '-fL', $tfvars_template_url{$args{deployment_type}}, '-o', $tfvars_file);
```
Comment (mpagot) [inline lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm:66]: what this culr is doing?

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+    assert_script_run($retrieve_tfvars_cmd);
+    assert_script_run("test -f $tfvars_file");
+    replace_tfvars_variables($tfvars_file);
+    upload_logs($tfvars_file, log_name => "$args{deployment_type}.tfvars.txt");
```
Comment (mpagot) [inline lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm:71]: `.txt` ;-) 

should we use the same trick in qesap tests?

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    replace_tfvars_variables();
+
+=over 1
+
+=item B<$deployment_type>: Type of the deployment (workload_zone, sap_system, library... etc)
+
+=back
+
+Replaces placeholder pattern B<%OPENQA_VARIABLE%> with corresponding OpenQA variable value.
+If OpenQA variable is not set, placeholder is replaced with empty value.
```
Comment (mpagot) [inline lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm:86]: this one is a big difference in comparision to what HanaSR is doing

File: lib/sles4sap/azure_cli.pm (hunk 4 above)
Comment (lpalovsky) [reply lib/sles4sap/azure_cli.pm:1458]: If az cli command fails to acquire a lease to the file, it will just return an error message which is not a valid json format. If you pass this through `decode_json()` as we usually do, the test will die with `malformed json format` error. 
The purpose of this function is to either return UUID or nothing instead of failing.

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm
Comment (mpagot) [reply lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm:1]: so code is not new but moved to a dedicated file

File: lib/sles4sap/azure_cli.pm (hunk 5 above)
Comment (lpalovsky) [reply lib/sles4sap/azure_cli.pm:1466]: The purpose of this function is to request exclusive file rights by calling `az storage blob lease acquire`. If call is successful, it will return UUID which you use afterwards for accessing the file. 

So if the function returns UUID, the file is yours. If it returns nothing, it means you won't get the rights and is up to the caller to decide how to deal with this situation - it can make the test fail completely, or retry in a loop. 

I just updated the description to:
 ```
Acquire a lease for a storage blob. Function returns UUID which is then required to modify the file and gives the
UUID owner exclusive rights.
Optionally B<lease_duration> can be defined to limit this file lock up to 60s instead of infinity.

```

File: lib/sles4sap/azure_cli.pm (hunk 2 above)
Comment (lpalovsky) [reply lib/sles4sap/azure_cli.pm:15]: Yup, agree.

File: lib/sles4sap/azure_cli.pm (hunk 9 above)
Comment (lpalovsky) [reply lib/sles4sap/azure_cli.pm:1512]: Actually... that is a good point... I wanted to use `assert_script_run` just to make it fail, but I'll leave it as it is and make the caller decide.  
I need to fix the other part of the code that calls this function.

Comment (lpalovsky) [reply lib/sles4sap/azure_cli.pm]: I just updated the desc.

File: lib/sles4sap/azure_cli.pm (hunk 8 above)
Comment (lpalovsky) [reply lib/sles4sap/azure_cli.pm]: Thanks, fixed

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm
Comment (lpalovsky) [reply lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm:1]: Yes. I moved the tfvars related functions into separate lib. 
I would not decouple those as it is related to filling in subnet variables.

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm (hunk 10 above)
Comment (lpalovsky) [reply lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm:66]: It retrieves the tfvars template from data if I remember correctly.

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm (hunk 11 above)
Comment (lpalovsky) [reply lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm:71]: I think I stole this one from qesap :D

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        library => 'library_parameter_file'
+    );
+    croak "Unknown deployment type: $args{deployment_type}" unless $tfvars_os_variable{$args{deployment_type}};
+
+    my %tfvars_template_url = (
+        deployer => data_url('sles4sap/sap_deployment_automation_framework/DEPLOYER.tfvars'),
+        sap_system => data_url('sles4sap/sap_deployment_automation_framework/SAP_SYSTEM.tfvars'),
+        workload_zone => data_url('sles4sap/sap_deployment_automation_framework/WORKLOAD_ZONE.tfvars'),
+        library => data_url('sles4sap/sap_deployment_automation_framework/LIBRARY.tfvars')
+    );
```
Comment (alvarocarvajald) [inline lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm:60]: A bit of Perl's crazyness suggestion:
```
my @valid_deployments = qw(deployer sap_system workload_zone library);
croak "Unknown deployment type: $args{deployment_type}" unless (grep /$args{deployment_type/ @valid_deployments);

my %tfvars_os_variable = map { $_ => "${_}_parameter_file" } @valid_deployments;
my %tfvars_template_url = map { $_ => data_url('sles4sap/sap_deployment_automation_framework/' . uc($_) . '.tfvars') } @valid_deployments;
```

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+    # replace default vnet name with shorter one to avoid naming restrictions
+    set_workload_vnet_name();
+
+    my $tfvars_file = get_os_variable($tfvars_os_variable{$args{deployment_type}});
+    my $retrieve_tfvars_cmd = join(' ', 'curl', '-v', '-fL', $tfvars_template_url{$args{deployment_type}}, '-o', $tfvars_file);
+
+    assert_script_run($retrieve_tfvars_cmd);
```
Comment (alvarocarvajald) [inline lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm:68]: Why not do directly:
```
assert_script_run("curl -v -fL $tfvars_template_url{$args{deployment_name}} -o $tfvars_file");
```
The line ends up shorter and you avoid creating a new variable.

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+Replaces placeholder pattern B<%OPENQA_VARIABLE%> with corresponding OpenQA variable value.
+If OpenQA variable is not set, placeholder is replaced with empty value.
+
+=cut
+
+sub replace_tfvars_variables {
+    my ($tfvars_file) = @_;
+    croak 'Variable "$tfvars_file" undefined' unless defined($tfvars_file);
+    my @variables = split("\n", script_output("grep -oP \'(\?<=%)[A-Z_]+(?=%)\' $tfvars_file"));
+    my %to_replace = map { '%' . $_ . '%' => get_var($_, '') } @variables;
```
Comment (alvarocarvajald) [inline lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm:94]: Is it OK that this can end up replacing some of the placeholders for empty strings if the settings are not defined?

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+Calculate network IP space that will be reserved to contain all underlying subnets: 'admin', 'db', 'app', 'web'.
+Each network space will reserve 64 IP addresses. This splits last IP octet into 4 network spaces.
+First 64 address spaces are used from third octet, making 256 network spaces available for all OpenQA tests at
+any point in time. (256 tests can run at the same time using one control plane.)
+
+=cut
+
+sub calculate_net_addr_space {
+    # map { $_ * 64 IP addresses } (1 .. all 256 addresses within octet/64 IP per network space);
+    my @addr_ranges = map { $_ * 64 } (1 .. 256 / 64);
```
Comment (alvarocarvajald) [inline lib/sles4sap/sap_deployment_automation_framework/networking.pm:44]: Why `1 .. 256 / 64` instead of `1 .. 4`? Or the intention is for this to return `1/64, 2/64, 3/64, .... 256/64`, because if it's the second, then it needs some parameters. `/` will take precedence over `..`

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+      app_subnet_address_prefix
+      web_subnet_address_prefix
+    );
+    my @octets = (split(/[.\/]/, $args{network_space}))[0 .. 3];
+    my @addr_ranges = map { $_ * 64 } (0 .. 3);
+    my $subnet_ip_count = 256 / @addr_ranges / @subnet_names;
+
+    my @subnet = map { join('.', @octets[0 .. 2], ($octets[3] + $subnet_ip_count * $_)) . '/28' } (0 .. 3);
+    my %network_data;
+    @network_data{@subnet_names} = @subnet;
```
Comment (alvarocarvajald) [inline lib/sles4sap/sap_deployment_automation_framework/networking.pm:82]: Oh! Didn't know that syntax was accepted. Neat.

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm [hunk 18]
```diff
@@ -0,0 +1,389 @@
...
+
+=cut
+
+sub acquire_network_file_lease {
+    my (%args) = @_;
+    $args{storage_account} //= get_required_var('SDAF_TFSTATE_STORAGE_ACCOUNT');
+    foreach ('network_lease_file', 'storage_account') {
+        croak "Missing mandatory argument: '$_'" unless $args{$_};
+    }
+    my $lease_id;
```
Comment (alvarocarvajald) [inline lib/sles4sap/sap_deployment_automation_framework/networking.pm:164]: I'd remove this line, and ....

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm [hunk 19]
```diff
@@ -0,0 +1,389 @@
...
+    my %arguments = (
+        container_name => 'network-spaces',
+        storage_account_name => $args{storage_account},
+        blob_name => $args{network_lease_file},
+        lease_duration => 60
+    );
+
+    # First acquire a file lease to gain exclusive file access - prevents other tests modifying it.
+    $lease_id = az_storage_blob_lease_acquire(%arguments);
```
Comment (alvarocarvajald) [inline lib/sles4sap/sap_deployment_automation_framework/networking.pm:173]: Add it here as:
```
my $lease_id = az_storage_blob_lease_acquire(%arguments);
```

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm [hunk 20]
```diff
@@ -0,0 +1,389 @@
...
+
+    # Return unless there file lease was successful
```
Comment (alvarocarvajald) [inline lib/sles4sap/sap_deployment_automation_framework/networking.pm:175]: s/there/the/

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm [hunk 21]
```diff
@@ -0,0 +1,389 @@
...
+    my $generated_address_space;
+    my $start_time = time();
+
+    record_info('New lease', 'All existing network files in usage: Creating new file');
+    while (!$address_space) {
+        die('Failed to reserve free address space within timeout') if (time() - $start_time > $args{timeout});
+        ($generated_address_space = calculate_net_addr_space()) =~ s/\/26//;
+        # Calculate another network if file for this network is already present
+        sleep 1;    # one sec delay to prevent command spamming
+        next if grep { $_ eq $generated_address_space } @{list_network_lease_files()};
```
Comment (alvarocarvajald) [inline lib/sles4sap/sap_deployment_automation_framework/networking.pm:348]: Doesn't `list_network_lease_files` returns a JSON object? Wondering if there may be problems by treating it as a list, or if the `query => [].name` prevents this.

File: lib/sles4sap/azure_cli.pm (hunk 3 above)
Comment (lpalovsky) [reply lib/sles4sap/azure_cli.pm:1399]: fixed

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm (hunk 13 above)
Comment (lpalovsky) [reply lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm:60]: As much as I like maps, I am inclined to keep more readable version here.
 
In any case... I have a very secret and very evil plan to quietly and slowly remove templates and generate full tfvars inside testcode. So those lines will just go away. :D
Don't tell anyone.

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm (hunk 14 above)
Comment (lpalovsky) [reply lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm]: True, makes sense. 
Fixed.

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm (hunk 16 above)
Comment (lpalovsky) [reply lib/sles4sap/sap_deployment_automation_framework/networking.pm]: It's the first one. The final array is `64, 128, 192,  256`. 
I was just trying to make it self explanatory how did I get to the number 4.
I'll change it to 4 since it has comment above explaining it .

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm (hunk 19 above)
Comment (lpalovsky) [reply lib/sles4sap/sap_deployment_automation_framework/networking.pm:173]: Yeah that makes sense. It might have been a remnant of an older code structure. 
Thanks. Fixed.

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm (hunk 21 above)
Comment (lpalovsky) [reply lib/sles4sap/sap_deployment_automation_framework/networking.pm:348]: It returns normal perl structure. 
`list_network_lease_files` returns output from `az_storage_blob_list`
and that is : `return decode_json(script_output($az_cmd));`
So it's already decoded.

**Comment by lpalovsky** [comment]
VRs with latest changes, both running in parallel:
https://openqaworker15.qa.suse.cz/tests/298401#live
https://openqaworker15.qa.suse.cz/tests/298403#live

**Review by mpagot — APPROVED** [review]
LGTM

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm (hunk 21 above)
Comment (alvarocarvajald) [reply lib/sles4sap/sap_deployment_automation_framework/networking.pm:347]: Yes, but `decode_json` returns a hash, not an array. That's my question. Not sure if it can cause problems.

**Review by alvarocarvajald — APPROVED** [review]
LGTM

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm [hunk 22]
```diff
@@ -0,0 +1,389 @@
...
+our @EXPORT = qw(
+  assign_address_space
+  calculate_subnets
+);
+
+=head2 calculate_net_addr_space
+
+    calculate_net_addr_space();
+
+Calculate network IP space that will be reserved to contain all underlying subnets: 'admin', 'db', 'app', 'web'.
```
[the remark this answers is not in the document]
Comment (lpalovsky) [reply lib/sles4sap/sap_deployment_automation_framework/networking.pm:35]: The randomness ... basically in whole code chunk has no functional impact there. 
It only decreases the chance of multiple tests spending time checking the same network as only one of those will succeed and the others have to go through the process again. If they do it in orderly manner everything will just pool up. 
Imagine having 4 tests and IP ranges A, B , C:
- Test 1 will pick IP range A, tests 2-4 will fail and go check IP B in the next round.  
- Test 2 will pick IP range B, tests 3-4 will fail and go for next loop
- Test 4 will have to go through 4 loops to finally get an IP instead of first one. 

Two random picks: 
- last octet is still part of whole address space and includes 4 networks for each bit in 3rd octet
- if I am correct, removing randomness will decrease entropy by 4. With 256 address spaces it is fine, but if we lower the number, chances of wasting time will go up.

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm [hunk 23]
```diff
@@ -0,0 +1,389 @@
...
+
+=over 2
+
+=item B<storage_account>: Storage account containing lease file. Default: 'SDAF_TFSTATE_STORAGE_ACCOUNT'
+
+=item B<network_lease_file>: Name of the lease file
+
+=back
+
+Acquire network lease for a blob.
```
[the remark this answers is not in the document]
Comment (lpalovsky) [reply lib/sles4sap/sap_deployment_automation_framework/networking.pm:157]: Done. 
```Acquire network lease for a blob. Returns blob lease UUID which is later required for getting permission to modify
blob file. ```

File: data/sles4sap/sap_deployment_automation_framework/WORKLOAD_ZONE.tfvars [hunk 24]
```diff
@@ -58,13 +58,13 @@ location = "%PUBLIC_CLOUD_REGION%"
...
 # The name is optional - it can be used to override the default naming
-#network_name = ""
+network_name = "%SDAF_SUT_VNET_NAME%"
 
 # network_arm_id is an optional parameter that if provided specifies Azure resource identifier for the existing Virtual Network
 #network_arm_id = ""
 
 # network_address_space is a mandatory parameter when an existing Virtual network is not used
-network_address_space = "10.10.0.0/16"
+network_address_space = "%NETWORK_ADDRESS_SPACE%"
```
[the remark this answers is not in the document]
Comment (lpalovsky) [reply data/sles4sap/sap_deployment_automation_framework/WORKLOAD_ZONE.tfvars:67]: Hmm... yes... the parameter is an internal one though. I will change all of them to comply with naming convention.

File: t/21_sles4sap_azure_cli.t [hunk 25]
```diff
@@ -750,4 +750,91 @@ subtest '[az_network_vnet_subnet_update]' => sub {
...
+    ok(grep(/--container-name Arlecchino/, @calls), 'Check for argument "--container-name"');
+    ok(grep(/--account-name Pantalone/, @calls), 'Check for argument "--account-name"');
+    ok(grep(/--file Colombina/, @calls), 'Check for argument "--file"');
+};
+
+subtest '[az_storage_blob_lease_acquire]' => sub {
+    my $azcli = Test::MockModule->new('sles4sap::azure_cli', no_auto => 1);
+    my @calls;
+    $azcli->redefine(script_output => sub { @calls = $_[0]; return '521fa121-4e04-448e-a8ec-d17e6b9c5e78'; });
+    $azcli->redefine(qesap_az_validate_uuid_pattern => sub { return '521fa121-4e04-448e-a8ec-d17e6b9c5e78'; });
```
[the remark this answers is not in the document]
Comment (lpalovsky) [reply t/21_sles4sap_azure_cli.t]: removed

File: lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm (hunk 15 above)
Comment (lpalovsky) [reply lib/sles4sap/sap_deployment_automation_framework/configure_tfvars.pm]: I created a new ticket to review whole tfvars mechanism: TEAM-9786
This was a function to provide a good enough solution.

File: lib/sles4sap/sap_deployment_automation_framework/networking.pm (hunk 21 above)
Comment (lpalovsky) [reply lib/sles4sap/sap_deployment_automation_framework/networking.pm:351]: No, because `query => [].name` specifies to return an array of names.

**Comment by lpalovsky** [comment]
Executing new VRs with suggestions applied
2 parallel tests: 
Test A: 
- finds unused lease file: https://openqaworker15.qa.suse.cz/tests/301361#step/deploy_workload_zone/77
- assigns address: https://openqaworker15.qa.suse.cz/tests/301361#step/deploy_workload_zone/98

Test B: 
- detects no free lease file: https://openqaworker15.qa.suse.cz/tests/301359#step/deploy_workload_zone/77
- creates new lease file and assigns address: https://openqaworker15.qa.suse.cz/tests/301359#step/deploy_workload_zone/98
